### PR TITLE
Support attribute macros

### LINKED ISSUE #1719: Listing fails when [attr] alias is used
This format works as expected:
~~~ ini
*.png filter=lfs diff=lfs merge=lfs -text
~~~
~~~ sh
> git lfs track
Listing tracked patterns
    *.png (.gitattributes)
~~~

But if I use an `[attr]` alias then `git lfs track` (list) on the command line fails to list anything…
~~~ ini
[attr]lfs filter=lfs diff=lfs merge=lfs -text
*.png   lfs
~~~
~~~ sh
> git lfs track
Listing tracked patterns
    [attr]lfs (.gitattributes)
~~~

### PR DESCRIPTION
**Pull request by bk2204**
It's possible in Git to use an attribute macro to specify a set of attributes and then apply them over a set of patterns like so:

```
[attr]lfs filter=lfs diff=lfs merge=lfs -text
*.png lfs
*.gif lfs
```

This series introduces some refactors so that we use one set of attribute parsing code and then implements macros on top of that newly refactored base.

As a note, we need to update `git-lfs/wildmatch`, since the previous version doesn't properly escape hash marks, leading to test failures (and in edge cases, breakage) when we standardize on using wildmatch for all patterns.

This series fixes #1719.

### CODE REVIEW

File: t/t-attributes.sh [hunk 1]
```diff
@@ -55,3 +55,26 @@ begin_test "macros with HOME"
...
+
+  printf '*.dat lfs\n' > .gitattributes
+  git add .gitattributes
+  git commit -m 'initial import'
+
+  contents="some data"
+  printf "$contents" > foo.dat
+  git add *.dat
+  git commit -m 'foo.dat'
+  assert_local_object "$(calc_oid "$contents")" 9
```
Comment (PastelMobileSuit) [inline t/t-attributes.sh]: Same here as above: I think we need a `git lfs track` call to ensure the pattern is already supported

File: t/t-attributes.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  contents2="other data"
+  printf "$contents" > dir/foo.bin
```
[the remark this answers is not in the document]
Comment (bk2204) [reply t/t-attributes.sh]: Ah, yeah, that should be `$contents2`.

File: t/t-attributes.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+(
+  set -e
+
+  reponame="$(basename "$0" ".sh")"
+  clone_repo "$reponame" repo
+
+  mkdir dir
+  printf '[attr]lfs filter=lfs diff=lfs merge=lfs -text\n*.dat lfs\n' \
+    > .gitattributes
+  printf '[attr]lfs2 filter=lfs diff=lfs merge=lfs -text\n*.dat lfs2\n' \
```
[the remark this answers is not in the document]
Comment (bk2204) [reply t/t-attributes.sh]: Yup.

File: t/t-attributes.sh [hunk 4]
```diff
@@ -36,3 +36,22 @@ begin_test "macros"
...
+
+  mkdir -p "$HOME/.config/git"
+  printf '[attr]lfs filter=lfs diff=lfs merge=lfs -text\n*.dat lfs\n' \
+    > "$HOME/.config/git/attributes"
+
+  contents="some data"
+  printf "$contents" > foo.dat
+  git add *.dat
+  git commit -m 'foo.dat'
+  assert_local_object "$(calc_oid "$contents")" 9
```
[the remark this answers is not in the document]
Comment (bk2204) [reply t/t-attributes.sh]: You're right.  Git will already have run it through the filter, so the only case in which we wouldn't have the object would be if Git hadn't honored the setting, and we're not testing Git.

I'll leave this as it is for the purposes of ensuring general sanity, but add a call to `git lfs track`.

[the remark this answers is not in the document]
Comment (bk2204) [reply t/t-attributes.sh]: Ah, one thing to mention is that on this line (unlike below), we're working solely with files outside of the working tree, which `git lfs track` intentionally ignores (except to read macros).  Below, we don't have that problem, since the file in the working tree isn't ignored.

I'll need to come up with a different way to test this part.

**Comment by bk2204** [comment]
Okay, all issues addressed.
